Fix reloading scripts already in use

Hilderin opened the pull request:
- Fixes #97138

Passing the parameter `p_keep_state` to `reload` in `Script::reload_from_file` fixed the issue.

But, I saw another problem when the scripts are opened in the editor, they were reloaded twice, once in the Script Editor and another time in EditorFileSystem. I added a check in `EditorFileSystem::_update_script_documentation` to reload script only if they are not opened in the Script Editor.

Hilderin committed:
Fix reloading scripts already in use

ryevdokimov commented:
Could you elaborate on the other problem? If I modify a single script externally the reload is still being fired multiple times based on a print_line() test, but maybe I'm misunderstanding something.

Also, not sure if you want to fix this here, but this comment from your related previous PR is confusing.

I'm assuming you meant: "The only reason the script is already loaded here is to allow editing it outside the editor, without being connected to the LSP server."

https://github.com/godotengine/godot/blob/9fd431f078d7560835da2642c9efe9e235618fe9/editor/editor_file_system.cpp#L1970-L1971

Hilderin commented:
> Could you elaborate on the other problem? If I modify a single script externally the reload is still being fired multiple times based on a print_line() test, but maybe I'm misunderstanding something.

The problem I fixed was related to using the internal editor. However, you are right — using an external editor connected to the LSP server (e.g., VSCode) was causing a double reload. If you edit the file in Notepad, for example, or use `git pull` to update a GDScript while VSCode is closed, there’s no LSP server to inform Godot that the script was updated. What can be tricky is that even if you update the file in Notepad while VSCode is open and connected via LSP, it’s the LSP server that will trigger the script reload. However, after some testing, it’s not always the case, and I couldn’t find a consistent pattern.

After further investigation, I found that scripts were already being reloaded twice when edited in an external editor connected via LSP — once from `GDScriptTextDocument::didSave` and once in `EditorNode::_resources_changed`. I made additional modifications to prevent reloading the script more than once when saved via an external editor, whether LSP is used or not. Using the `get_last_modified_time` of the resource and comparing it with the file’s last modification time on disk in `EditorFileSystem::_should_reload_script` and `EditorNode::_resources_changed` solved the issue.

Additionally, I discovered that both the ScriptEditor and LSP server use `get_language()->reload_tool_script` to reload tool scripts. I couldn't see any difference between that and a simple `reload(true)`, but just in case, I modified `Script::reload_from_file` to do exactly the same thing.

Note: The `GDScript::reload` method is still called twice when invoking `Script::reload_from_file` — once in `GDScriptCache::get_full_script` and once directly in `Script::reload_from_file`. This behavior existed before, and I didn’t change it.

> I'm assuming you meant: "The only reason the script is already loaded here is to allow editing it outside the editor, without being connected to the LSP server."

Indeed, good catch. However, I removed that comment in the new version and created a function to check if the script needs to be reloaded: `EditorFileSystem::_should_reload_script`.

---

I also found a bug where adding or removing exported properties in tool scripts doesn't update live when the script is saved. You need to select another node and then reselect the node to refresh the property list in the inspector. I tested this without the PR, and the same thing happens. I'll create a separate issue for that.

Hilderin commented on editor/editor_file_system.cpp:
```diff
@@ -44,6 +44,7 @@
 #include "editor/editor_settings.h"
 #include "editor/plugins/script_editor_plugin.h"
 #include "editor/project_settings_editor.h"
+#include "modules/gdscript/language_server/gdscript_language_protocol.h"
```
You are right, it was only à test, I'll remove it.

Hilderin replied:
Done!

akien-mga commented on editor/editor_file_system.cpp:
```diff
@@ -2006,6 +2004,36 @@ void EditorFileSystem::_update_script_documentation() {
+		return false;
+	}
+
+	if (res->get_last_modified_time() == FileAccess::get_modified_time(p_path)) {
+		// Already up to date.
+		return false;
+	}
```
I wonder if this can cause issues for the Android editor like #94416.
If the implementation for `get_modified_time` is still buggy on Android, this might be problematic. https://github.com/godotengine/godot/issues/94416#issuecomment-2267172619

Hilderin replied:
You are so right. I'm trying to find a way to prevent script double/triple script reloading from file.

After some more testing, I removed the `reloading_from_file` completely for script in `EditorNode`. I did not find a way to scripts not being reloaded from file now that EditorFileSystem do it in `EditorFileSystem::_update_script_documentation`. That way, no need to check the last modified time and not hack in `EditorNode` for scripts.

Technically, the method `EditorFileSystem::_update_script_documentation` should be renamed now for  `EditorFileSystem::_update_scripts` but I'm not sure it's a good idea to rename it in this PR. What do you think?

KoBeWi commented:
Not sure if related, but there is a bug where trying to run a build-in tool script (with File -> Run) will make it completely unable to be modified until editor restart (even closing and re-opening the scene will not help). It prints the same "already in use error").

Example:
```GDSCript
@tool
extends Node

func _process(delta: float) -> void:
	print("A")
```
It will keep printing A. Change it to B and it will print B. Press Ctrl+Shift+X, you will get error in the console and any further script modifications will be ignored.

I can open a new issue, but since it involves the same error, I though you could fix it too.

Hilderin commented:
It's definitively related, once you get the `already in use error` the script will never reload after that. I'll try your example.

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -2006,6 +2004,31 @@ void EditorFileSystem::_update_script_documentation() {
+		return false;
+	}
+
+	Ref<Resource> res = ResourceCache::get_ref(p_path);
+	if (res.is_null()) {
+		// Script not already loaded.
+		return false;
+	}
+
+	Ref<Script> scr = res;
+	if (scr.is_null()) {
+		// Not a script.
+		return false;
+	}
```
You can merge these 2 checks.
```suggestion
	Ref<Script> scr = ResourceCache::get_ref(p_path);
	if (scr.is_null()) {
		// Not a script.
		return false;
	}
```

Hilderin replied:
Done!

Hilderin commented:
> It will keep printing A. Change it to B and it will print B. Press Ctrl+Shift+X, you will get error in the console and any further script modifications will be ignored.

I tested your script with this PR and it fixes the issue. I can update the script inside Godot, External script with LSP or not and it's updating perfectly!

Hilderin commented:
> It doesn't for me.

My bad, I tested it too quickly. The issue you're describing existed in 4.3 stable when trying to run a tool script that was still in memory. It's the same problem: before running the script, a `reload(false)` was executed, causing the same error message. Simply replacing `false` with `true` fixed the issue. Everywhere else in the `ScriptEditor`, `reload` was called with `true`.

I also retested with an `EditorScript`, and it still works fine.

I searched through the codebase, and the only places where `reload` could be called with `false` are in `GDScriptLanguage::reload_scripts` and `GDScriptLanguage::reload_tool_script`. I assume these could be called outside the editor?! I'm still not sure why we would want to pass `p_soft_reload` as `false`.

akien-mga commented:
Thanks!

dmlary commented:
Is there any chance we can get this fix cherry-picked for 4.3.1?  It's painful to work on tool scripts without it right now on 4.3.

akien-mga commented:
This is a fix for a regression specific to 4.4, I don't think it's applicable as is to 4.3. Do you have a bug report detailing what issue you have in 4.3?

dmlary commented:
> This is a fix for a regression specific to 4.4, I don't think it's applicable as is to 4.3. Do you have a bug report detailing what issue you have in 4.3?

This same error and behavior happens in 4.3 for `@tool` scripts edited in an external editor that add menus or connect to signals. The error appears during the first hot reload, but the changes are applied. Any subsequent edit will not hot reload. 

I can open a separate issue to document it if necessary.

dmlary commented:
Opened #97280 to reflect a similar problem in 4.3.
